File: 01975. Maximum Matrix Sum/solution.ts
/**
 * Problem: 1975. Maximum Matrix Sum
 *
 * Difficulty: Medium
 *
 * Language: TypeScript
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

/**
 * Calculate the maximum possible sum after flipping any number of signs
 *
 * @param matrix - 2D array of integers
 *
 * @returns - Maximum possible sum after optimal flips
 */
const maxMatrixSum = (matrix: number[][]): number => {
     // Sum of absolute values of all elements
    let totalSum = 0
    // Minimum absolute value in the matrix
    let minAbs = Infinity
    // Count of negative numbers
    let negativeCount = 0

    // Iterate through all elements in the matrix
    for (let i = 0; i < matrix.length; i++) {
        for (let j = 0; j < matrix[i].length; j++) {
            const val = matrix[i][j]
            
            // Count negative numbers
            if (val < 0) negativeCount++

            // Add absolute value to total sum
            totalSum += Math.abs(val)
            // Update minimum absolute value
            minAbs = Math.min(minAbs, Math.abs(val))
        }
    }

    // If even number of negatives, we can make all values positive
    if (negativeCount % 2 === 0) return totalSum

    // If odd number of negatives, subtract twice the minimum absolute value
    // (flip the smallest absolute value to negative to minimize loss)
    return totalSum - 2 * minAbs
}

File: 01975. Maximum Matrix Sum/solution.cs
/**
 * Problem: 1975. Maximum Matrix Sum
 *
 * Difficulty: Medium
 *
 * Language: C#
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

using System.Runtime.CompilerServices;

public class Solution {
    // Apply aggressive optimization hints to the compiler
    [MethodImpl(MethodImplOptions.AggressiveInlining | MethodImplOptions.AggressiveOptimization)]
    public long MaxMatrixSum(int[][] matrix) {
        // Initialize variables
        var min = int.MaxValue;    // Minimum absolute value found
        var sum = 0L;              // Total sum of absolute values (long to prevent overflow)
        var xor = 0;               // XOR of all original values (bitwise trick to track parity)
        
        // Iterate through all elements in the matrix
        foreach (var row in matrix) {
            foreach (var n in row) {
                // Calculate absolute value using bitwise operations:
                // sign = n >> 31 gives -1 for negative numbers, 0 for non-negative
                // (n ^ sign) - sign computes absolute value efficiently
                var sign = n >> 31;
                var abs = (n ^ sign) - sign;
                
                // Accumulate sum of absolute values
                sum += abs;
                // XOR all original values to track parity of negative count
                xor ^= n;
                // Update minimum absolute value
                min = Math.Min(min, abs);
            }
        }
        
        // Calculate final result:
        // - If xor >> 31 is 1 (odd number of negatives), subtract 2*min
        // - If xor >> 31 is 0 (even number of negatives), subtract 0
        return sum - (min << 1 & xor >> 31);
    }
}
